File: spring-backend/src/main/java/com/simmortal/shop/ShopController.java
package com.simmortal.shop;

import com.simmortal.memorial.MemorialDecoration;
import com.simmortal.memorial.MemorialDonationWreath;
import com.simmortal.memorial.MemorialTribute;
import java.util.Map;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestHeader;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("shop")
public class ShopController {
  private final ShopService shopService;

  public ShopController(ShopService shopService) {
    this.shopService = shopService;
  }

  @GetMapping
  public Object getShopContent(@RequestParam(value = "locale", required = false) String locale) {
    return shopService.getShopContent(locale);
  }

  @GetMapping("/{slug}")
  public Object getShopItemBySlug(
      @PathVariable String slug,
      @RequestParam(value = "locale", required = false) String locale,
      @RequestHeader(value = "X-User-Id", required = false) String userId
  ) {
    return shopService.getShopItemBySlug(slug, locale, userId);
  }

  @PostMapping("webhook")
  public void handleWebhookEvent(@RequestBody Map<String, Object> event) {
    Object typeValue = event.get("type");
    String type = typeValue == null ? "" : typeValue.toString();
    if ("customer.subscription.created".equals(type)) {
      shopService.handleSubscriptionCreated(event);
    }
    if ("customer.subscription.updated".equals(type) || "customer.subscription.deleted".equals(type)) {
      shopService.handleSubscriptionUpdated(event);
    } else if ("payment_intent.succeeded".equals(type)) {
      shopService.handlePaymentIntentSucceeded(event);
    }
    if ("invoice.payment_succeeded".equals(type)) {
      shopService.handleInvoicePaymentSucceeded(event);
    }
  }

  @PostMapping("order")
  public CreateOrderResponse createOrder(
      @RequestBody CreateOrderRequest request,
      @RequestHeader("X-User-Id") String userId
  ) {
    return shopService.createOrder(userId, request);
  }

  @GetMapping("order/{orderId}")
  public Object getOrderById(
      @PathVariable String orderId,
      @RequestParam("status") OrderStatus status,
      @RequestHeader("X-User-Id") String userId
  ) {
    return shopService.getOrderById(userId, orderId, status);
  }

  @PostMapping("waitlist/{itemId}")
  public void joinWaitlist(
      @PathVariable String itemId,
      @RequestHeader("X-User-Id") String userId
  ) {
    shopService.joinWaitlist(userId, itemId);
  }

  @GetMapping("price/decoration")
  public Map<MemorialDecoration, String> getDecorationPrices() {
    return shopService.getAllDecorationPrices();
  }

  @GetMapping("price/tribute")
  public Map<MemorialTribute, String> getTributePrices() {
    return shopService.getAllTributePrices();
  }

  @GetMapping("price/wreath")
  public Map<MemorialDonationWreath, String> getWreathPrices() {
    return shopService.getAllWreathPrices();
  }
}

File: Backend-main/src/shop/shop.controller.ts
import { Body, Controller, Get, Param, Post, Query, UseInterceptors } from '@nestjs/common';
import { ShopListingsResponse, ShopItem, Order, OrderStatus } from './interface/shop.interface';
import { PublicAccess, Session } from 'supertokens-nestjs';
import { ShopService } from './shop.service';
import Stripe from 'stripe';
import { CacheInterceptor, CacheKey, CacheTTL } from '@nestjs/cache-manager';
import { CreateOrderRequest, CreateOrderResponse } from './interface/shop.dto';
import { SessionContainer } from 'supertokens-node/recipe/session';
import {
  MemorialDecoration,
  MemorialDonationWreath,
  MemorialTribute,
} from 'src/memorial/interface/memorial.interface';
import { PublicWithOptionalSession } from 'src/guard/optional-session.guard';

@Controller('shop')
export class ShopController {
  constructor(private readonly shopService: ShopService) {}

  @PublicAccess()
  @Get()
  @UseInterceptors(CacheInterceptor)
  @CacheTTL(300000) // Cache for 5 minutes
  async getShopContent(@Query('locale') locale?: string): Promise<ShopListingsResponse> {
    return this.shopService.getShopContent(locale);
  }

  @PublicWithOptionalSession()
  @Get('/:slug')
  async getShopItemBySlug(
    @Param('slug') slug: string,
    @Query('locale') locale?: string,
    @Session() session?: SessionContainer,
  ): Promise<ShopItem> {
    return this.shopService.getShopItemBySlug(slug, locale, session?.getUserId());
  }

  @PublicAccess()
  @Post('webhook')
  async handleWebhookEvent(@Body() event: Stripe.Event): Promise<void> {
    if (event.type === 'customer.subscription.created') {
      await this.shopService.handleSubscriptionCreated(event);
    }
    if (
      event.type === 'customer.subscription.updated' ||
      event.type === 'customer.subscription.deleted'
    ) {
      await this.shopService.handleSubscriptionUpdated(event);
    } else if (event.type === 'payment_intent.succeeded') {
      await this.shopService.handlePaymentIntentSucceeded(event);
    }
    if (event.type === 'invoice.payment_succeeded') {
      await this.shopService.handleInvoicePaymentSucceeded(event);
    }
  }

  @Post('order')
  async createOrder(
    @Body() request: CreateOrderRequest,
    @Session() session: SessionContainer,
  ): Promise<CreateOrderResponse> {
    return this.shopService.createOrder(session.getUserId(), request);
  }

  @Get('order/:orderId')
  async getOrderById(
    @Param('orderId') orderId: string,
    @Query('status') status: OrderStatus,
    @Session() session: SessionContainer,
  ): Promise<Order> {
    return this.shopService.getOrderById(session.getUserId(), orderId, status);
  }

  @Post('waitlist/:itemId')
  async joinWaitlist(
    @Param('itemId') itemId: string,
    @Session() session: SessionContainer,
  ): Promise<void> {
    return this.shopService.joinWaitlist(session.getUserId(), itemId);
  }

  @Get('price/decoration')
  @UseInterceptors(CacheInterceptor)
  @CacheKey('decoration-prices')
  @CacheTTL(60000) // Cache for 1 minute
  async getDecorationPrices(): Promise<Record<MemorialDecoration, string>> {
    return this.shopService.getAllDecorationPrices();
  }

  @Get('price/tribute')
  @UseInterceptors(CacheInterceptor)
  @CacheKey('tribute-prices')
  @CacheTTL(60000) // Cache for 1 minute
  async getTributePrices(): Promise<Record<MemorialTribute, string | null>> {
    return this.shopService.getAllTributePrices();
  }

  @PublicAccess()
  @Get('price/wreath')
  @UseInterceptors(CacheInterceptor)
  @CacheKey('wreath-prices')
  @CacheTTL(60000) // Cache for 1 minute
  async getWreathPrices(): Promise<Record<MemorialDonationWreath, string>> {
    return this.shopService.getAllWreathPrices();
  }
}
